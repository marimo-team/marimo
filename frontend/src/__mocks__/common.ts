/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable no-console */
import { type Mock, vi } from "vitest";
import { invariant } from "@/utils/invariant";

// Common mock factories
export const Mocks = {
  quietLogger: () => ({
    debug: vi.fn(),
    log: vi.fn(),
    warn: vi.fn(),
    error: vi.fn(),
    trace: vi.fn(),
    get: vi.fn().mockImplementation(() => Mocks.quietLogger()),
    disabled: vi.fn(),
  }),

  logger: () => ({
    debug: vi.fn().mockImplementation(console.debug),
    log: vi.fn().mockImplementation(console.log),
    warn: vi.fn().mockImplementation(console.warn),
    error: vi.fn().mockImplementation(console.error),
    trace: vi.fn().mockImplementation(console.trace),
    get: vi.fn().mockImplementation(() => Mocks.logger()),
    disabled: vi.fn(),
  }),

  toast: vi.fn(),

  clipboard: () => ({
    write: vi.fn(),
    writeText: vi.fn(),
    read: vi.fn(),
    readText: vi.fn(),
  }),

  clipboardItem: (data: Record<string, string>) => ({
    types: Object.keys(data),
    getType: vi
      .fn()
      .mockImplementation((type: string) =>
        Promise.resolve(new Blob([data[type]?.toString() || ""], { type })),
      ),
    supports: vi.fn().mockReturnValue(true),
  }),

  blob: (parts?: string[], options?: { type?: string }) => ({
    text: () => Promise.resolve(parts?.[0] || ""),
    type: options?.type || "text/plain",
  }),

  event: <T extends React.SyntheticEvent>(props: Partial<T> = {}) => ({
    preventDefault: vi.fn(),
    continuePropagation: vi.fn(),
    target: document.createElement("div"),
    currentTarget: document.createElement("div"),
    stopPropagation: vi.fn(),
    ...props,
  }),

  keyboardEvent: (
    props: Partial<React.KeyboardEvent> = {},
  ): React.KeyboardEvent<HTMLElement> & {
    continuePropagation: () => void;
  } =>
    Mocks.event({
      ...props,
    }) as unknown as React.KeyboardEvent<HTMLElement> & {
      continuePropagation: () => void;
    },
};

// Global mock setup functions
export const SetupMocks = {
  clipboard: (mockClipboard = Mocks.clipboard()) => {
    Object.defineProperty(navigator, "clipboard", {
      value: mockClipboard,
      writable: true,
    });

    // Mock ClipboardItem
    // @ts-expect-error - ClipboardItem types not exact
    global.ClipboardItem = vi
      .fn()
      .mockImplementation((data) => Mocks.clipboardItem(data));

    global.ClipboardItem.supports = vi.fn().mockReturnValue(true);

    // Mock Blob
    global.Blob = vi
      .fn()
      .mockImplementation((parts, options) => Mocks.blob(parts, options));

    return mockClipboard;
  },

  localStorage: () => {
    const store: Record<string, string> = {};

    Object.defineProperty(window, "localStorage", {
      value: {
        getItem: vi.fn((key: string) => store[key] || null),
        setItem: vi.fn((key: string, value: string) => {
          store[key] = value;
        }),
        removeItem: vi.fn((key: string) => {
          // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
          delete store[key];
        }),
        clear: vi.fn(() => {
          // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
          Object.keys(store).forEach((key) => delete store[key]);
        }),
        key: vi.fn((index: number) => Object.keys(store)[index] || null),
        get length() {
          return Object.keys(store).length;
        },
      },
      writable: true,
    });
  },

  fetch: (mockImplementation?: Mock) => {
    const mockFetch = vi.fn(mockImplementation);
    global.fetch = mockFetch;
    return mockFetch;
  },

  websocket: () => {
    const mockWs = {
      close: vi.fn(),
      send: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      onopen: null,
      onclose: null,
      onmessage: null,
      onerror: null,
      readyState: WebSocket.CONNECTING,
    };

    global.WebSocket = vi.fn(() => mockWs) as unknown as typeof WebSocket;
    return mockWs;
  },
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function asMock<T extends (...args: any[]) => unknown>(fn: T): Mock<T> {
  invariant(
    "mock" in fn,
    "fn must be a mock function, use vi.fn() to create one",
  );
  return fn as unknown as Mock<T>;
}

export function partialImplementation<T extends object>(
  partial: Partial<T>,
): T {
  return partial as unknown as T;
}
