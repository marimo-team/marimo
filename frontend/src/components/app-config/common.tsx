/* Copyright 2024 Marimo. All rights reserved. */
import { PropsWithChildren } from "react";

export const SettingTitle: React.FC<PropsWithChildren> = ({ children }) => {
  return (
    <div className="text-md font-semibold text-muted-foreground uppercase tracking-wide  mb-1">
      {children}
    </div>
  );
};

export const SettingSubtitle: React.FC<PropsWithChildren> = ({ children }) => {
  return (
    <div className="text-sm font-semibold underline-offset-2 text-accent-foreground uppercase tracking-wide">
      {children}
    </div>
  );
};

export const SettingDescription: React.FC<PropsWithChildren> = ({
  children,
}) => {
  return <p className="text-sm text-muted-foreground">{children}</p>;
};
