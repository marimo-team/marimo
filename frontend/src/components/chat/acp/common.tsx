/* Copyright 2024 Marimo. All rights reserved. */

import type { RequestPermissionResponse } from "@zed-industries/agent-client-protocol";
import {
  CheckCircleIcon,
  Loader2,
  PlugIcon,
  ShieldCheckIcon,
  WifiIcon,
  WifiOffIcon,
  XCircleIcon,
} from "lucide-react";
import React, { memo } from "react";
import {
  Accordion,
  AccordionContent,
  AccordionItem,
  AccordionTrigger,
} from "@/components/ui/accordion";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import { cn } from "@/utils/cn";
import { ToolBodyBlock } from "./blocks";
import type { AgentPendingPermission } from "./types";

interface SimpleAccordionProps {
  title: string | React.ReactNode;
  children: React.ReactNode | string;
  status?: "loading" | "error" | "success";
  index?: number;
  defaultIcon?: React.ReactNode;
}

export const SimpleAccordion: React.FC<SimpleAccordionProps> = ({
  title,
  children,
  status,
  index = 0,
  defaultIcon,
}) => {
  const getStatusIcon = () => {
    switch (status) {
      case "loading":
        return <Loader2 className="h-3 w-3 animate-spin" />;
      case "error":
        return <XCircleIcon className="h-3 w-3 text-destructive" />;
      case "success":
        return <CheckCircleIcon className="h-3 w-3 text-[var(--blue-9)]" />;
      default:
        return defaultIcon;
    }
  };

  return (
    <Accordion
      key={`tool-${index}`}
      type="single"
      collapsible={true}
      className="w-full"
    >
      <AccordionItem value="tool-call" className="border-0">
        <AccordionTrigger
          className={cn(
            "py-1 text-xs border-border shadow-none! ring-0! bg-muted hover:bg-muted/30 px-2 gap-1 rounded-sm [&[data-state=open]>svg]:rotate-180",
            status === "error" && "text-destructive/80",
            status === "success" && "text-[var(--blue-8)]",
          )}
        >
          <span className="flex items-center gap-1">
            {getStatusIcon()}
            <code className="font-mono text-xs">{title}</code>
          </span>
        </AccordionTrigger>
        <AccordionContent className="p-2">
          <div className="space-y-3 max-h-64 overflow-y-auto scrollbar-thin">
            {status !== "error" && (
              <div className="text-xs font-medium text-muted-foreground mb-1">
                {children}
              </div>
            )}

            {status === "error" && (
              <div className="bg-destructive/10 border border-destructive/20 rounded-md p-3 text-xs text-destructive">
                {children}
              </div>
            )}
          </div>
        </AccordionContent>
      </AccordionItem>
    </Accordion>
  );
};

interface ConnectionStatusProps {
  status: string;
  className?: string;
}

export const ConnectionStatus: React.FC<ConnectionStatusProps> = memo(
  ({ status, className }) => {
    const getStatusConfig = () => {
      switch (status) {
        case "connected":
          return {
            icon: <WifiIcon className="h-3 w-3" />,
            label: "Connected",
            variant: "default" as const,
            className:
              "bg-[var(--blue-3)] text-[var(--blue-11)] border-[var(--blue-5)]",
          };
        case "connecting":
          return {
            icon: <PlugIcon className="h-3 w-3 animate-pulse" />,
            label: "Connecting",
            variant: "secondary" as const,
            className:
              "bg-[var(--yellow-3)] text-[var(--yellow-11)] border-[var(--yellow-5)]",
          };
        case "disconnected":
          return {
            icon: <WifiOffIcon className="h-3 w-3" />,
            label: "Disconnected",
            variant: "outline" as const,
            className:
              "bg-[var(--red-3)] text-[var(--red-11)] border-[var(--red-5)]",
          };
        default:
          return {
            icon: <WifiOffIcon className="h-3 w-3" />,
            label: status || "Unknown",
            variant: "outline" as const,
            className:
              "bg-[var(--gray-3)] text-[var(--gray-11)] border-[var(--gray-5)]",
          };
      }
    };

    const config = getStatusConfig();

    return (
      <Badge
        variant={config.variant}
        className={cn(config.className, className)}
      >
        {config.icon}
        <span className="ml-1 text-xs font-medium">{config.label}</span>
      </Badge>
    );
  },
);
ConnectionStatus.displayName = "ConnectionStatus";

interface PermissionRequestProps {
  permission: NonNullable<AgentPendingPermission>;
  onResolve: (option: RequestPermissionResponse) => void;
}

export const PermissionRequest: React.FC<PermissionRequestProps> = memo(
  ({ permission, onResolve }) => {
    return (
      <div className="border border-[var(--amber-8)] bg-[var(--amber-2)] rounded-lg p-2">
        <div className="flex items-center gap-2 mb-3">
          <ShieldCheckIcon className="h-4 w-4 text-[var(--amber-11)]" />
          <h3 className="text-sm font-medium text-[var(--amber-11)]">
            Permission Request
          </h3>
        </div>
        <p className="text-sm text-[var(--amber-11)] mb-3">
          The AI agent is requesting permission to proceed:
        </p>
        <ToolBodyBlock data={permission.toolCall} />
        <div className="flex gap-2">
          {permission.options.map((option) => (
            <Button
              key={option.optionId}
              size="xs"
              variant="text"
              className={
                option.kind.startsWith("allow")
                  ? "text-[var(--blue-10)]"
                  : "text-[var(--red-10)]"
              }
              onClick={() =>
                onResolve({
                  outcome: {
                    outcome: "selected",
                    optionId: option.optionId,
                  },
                })
              }
            >
              {option.kind.startsWith("allow") && (
                <CheckCircleIcon className="h-3 w-3 mr-1" />
              )}
              {option.kind.startsWith("reject") && (
                <XCircleIcon className="h-3 w-3 mr-1" />
              )}
              {option.name}
            </Button>
          ))}
        </div>
      </div>
    );
  },
);
PermissionRequest.displayName = "PermissionRequest";
