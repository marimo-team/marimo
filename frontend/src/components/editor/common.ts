/* Copyright 2024 Marimo. All rights reserved. */

import type { CellId } from "@/core/cells/ids";
import { HTMLCellId } from "@/core/cells/ids";

/**
 * Create DOM properties for a cell.
 *
 * Returns:
 * - `data-cell-id` - The ID of the cell.
 * - `data-cell-name` - The name of the cell.
 * - `id` - The ID of the cell.
 */
export function cellDomProps(cellId: CellId, cellName: string) {
  return {
    "data-cell-id": cellId,
    "data-cell-name": cellName,
    id: HTMLCellId.create(cellId),
  };
}
