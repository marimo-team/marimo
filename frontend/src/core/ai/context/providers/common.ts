/* Copyright 2024 Marimo. All rights reserved. */

export const Boosts = {
  LOCAL_TABLE: 5,
  REMOTE_TABLE: 4,
  VARIABLE: 3,
} as const;
