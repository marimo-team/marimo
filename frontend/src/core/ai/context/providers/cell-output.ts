/* Copyright 2024 Marimo. All rights reserved. */

import type { Completion } from "@codemirror/autocomplete";
import type { FileUIPart } from "ai";
import { toPng } from "html-to-image";
import { type NotebookState, notebookAtom } from "@/core/cells/cells";
import { type CellId, CellOutputId } from "@/core/cells/ids";
import { displayCellName } from "@/core/cells/names";
import { isOutputEmpty } from "@/core/cells/outputs";
import type { OutputMessage } from "@/core/kernel/messages";
import type { JotaiStore } from "@/core/state/jotai";
import { parseHtmlContent } from "@/utils/dom";
import { Logger } from "@/utils/Logger";
import { type AIContextItem, AIContextProvider } from "../registry";
import { contextToXml } from "../utils";
import { Boosts, Sections } from "./common";

export interface BaseOutput {
  processedContent?: string;
  imageUrl?: string;
  shouldDownloadImage?: boolean;
  output: OutputMessage;
  outputType: "text" | "media";
}

export interface CellOutputData {
  cellId: CellId;
  cellName: string;
  cellCode: string;
  cellOutput?: BaseOutput;
  consoleOutputs?: BaseOutput[];
}

// For the context provider
// Currently, we enforce that cellOutput is present.
interface CellOutputContextData
  extends CellOutputData,
    Record<string, unknown> {
  cellOutput: BaseOutput;
}

export interface CellOutputContextItem extends AIContextItem {
  type: "cell-output";
  data: CellOutputContextData;
}

function isMediaMimetype(
  mimetype: OutputMessage["mimetype"] | undefined,
  htmlString: string,
): boolean {
  if (!mimetype) {
    return false;
  }

  const mediaPrefixes = ["image/", "video/", "audio/", "application/pdf"];
  if (mediaPrefixes.some((prefix) => mimetype.startsWith(prefix))) {
    return true;
  }
  const mediaIncludes = ["svg", "vega"];
  if (mediaIncludes.some((include) => mimetype.includes(include))) {
    return true;
  }

  // If it is HTML, we need to check if it contains a media tag
  if (mimetype === "text/html") {
    const mediaTags = [
      "<img",
      "<video",
      "<audio",
      "<iframe",
      "<canvas",
      "<svg",
      "<marimo-ui-element",
    ];
    if (mediaTags.some((tag) => htmlString.includes(tag))) {
      return true;
    }
  }

  return false;
}

export class CellOutputContextProvider extends AIContextProvider<CellOutputContextItem> {
  readonly title = "Cell Outputs";
  readonly mentionPrefix = "@";
  readonly contextType = "cell-output";
  private store: JotaiStore;
  constructor(store: JotaiStore) {
    super();
    this.store = store;
  }

  getItems(): CellOutputContextItem[] {
    const notebook = this.store.get(notebookAtom);
    const items: CellOutputContextItem[] = [];

    for (const cellId of notebook.cellIds.inOrderIds) {
      const cellContextData = getCellContextData(cellId, notebook);
      const cellOutput = cellContextData.cellOutput;

      // Skip cells with no output
      if (!cellOutput) {
        continue;
      }

      const mimetype = cellOutput.output.mimetype || "unknown";

      items.push({
        uri: this.asURI(cellId),
        name: cellContextData.cellName,
        type: this.contextType,
        description: `Cell output (${mimetype})`,
        data: { ...cellContextData, cellOutput },
      });
    }

    return items;
  }

  formatCompletion(item: CellOutputContextItem): Completion {
    const { cellOutput: data, cellName, cellCode } = item.data;

    return {
      label: `@${cellName}`,
      displayLabel: cellName,
      detail: `${data.outputType} output`,
      boost: data.outputType === "media" ? Boosts.HIGH : Boosts.MEDIUM,
      type: this.contextType,
      section: Sections.CELL_OUTPUT,
      apply: `@${cellName}`,
      info: () => {
        const infoContainer = document.createElement("div");
        infoContainer.classList.add(
          "mo-cm-tooltip",
          "docs-documentation",
          "min-w-[300px]",
          "max-w-[500px]",
          "flex",
          "flex-col",
          "gap-2",
        );

        const headerDiv = document.createElement("div");
        headerDiv.classList.add("flex", "flex-col", "gap-1");

        const nameDiv = document.createElement("div");
        nameDiv.classList.add("font-bold", "text-base");
        nameDiv.textContent = cellName;
        headerDiv.append(nameDiv);

        const descriptionDiv = document.createElement("div");
        descriptionDiv.classList.add("text-sm", "text-muted-foreground");
        headerDiv.append(descriptionDiv);

        infoContainer.append(headerDiv);

        // Show cell code preview
        if (cellCode) {
          const codeHeaderDiv = document.createElement("div");
          codeHeaderDiv.classList.add(
            "text-xs",
            "font-medium",
            "text-muted-foreground",
          );
          codeHeaderDiv.textContent = "Code:";
          infoContainer.append(codeHeaderDiv);

          const codeDiv = document.createElement("div");
          codeDiv.classList.add(
            "text-xs",
            "font-mono",
            "bg-muted",
            "p-2",
            "rounded",
            "max-h-20",
            "overflow-y-auto",
          );
          codeDiv.textContent =
            cellCode.slice(0, 200) + (cellCode.length > 200 ? "..." : "");
          infoContainer.append(codeDiv);
        }

        // Show output preview
        if (data.processedContent) {
          const outputHeaderDiv = document.createElement("div");
          outputHeaderDiv.classList.add(
            "text-xs",
            "font-medium",
            "text-muted-foreground",
            "mt-2",
          );
          outputHeaderDiv.textContent = "Output Preview:";
          infoContainer.append(outputHeaderDiv);

          const outputDiv = document.createElement("div");
          outputDiv.classList.add(
            "text-xs",
            "bg-muted",
            "p-2",
            "rounded",
            "max-h-24",
            "overflow-y-auto",
            "mb-2",
          );
          outputDiv.textContent =
            data.processedContent.slice(0, 300) +
            (data.processedContent.length > 300 ? "..." : "");
          infoContainer.append(outputDiv);
        }

        if (data.outputType === "media") {
          const mediaDiv = document.createElement("div");
          mediaDiv.classList.add(
            "text-xs",
            "text-muted-foreground",
            "italic",
            "mb-2",
          );
          mediaDiv.textContent = "A screenshot of the output will be attached";
          infoContainer.append(mediaDiv);
        }

        return infoContainer;
      },
    };
  }

  formatContext(item: CellOutputContextItem): string {
    const { cellOutput: data, cellName, cellId, cellCode } = item.data;

    const contextData = {
      name: cellName,
      cellId: cellId,
      outputType: data.outputType,
      mimetype: data.output.mimetype,
    } as const;

    let details = `Cell Code:\n${cellCode}\n\n`;

    if (data.outputType === "text" && data.processedContent) {
      details += `Output:\n${data.processedContent}`;
    } else if (data.outputType === "media") {
      details += `Media Output: Contains ${data.output.mimetype} content`;
      if (data.imageUrl) {
        details += `\nImage URL: ${data.imageUrl}`;
      }
    }

    return contextToXml({
      type: this.contextType,
      data: contextData,
      details,
    });
  }

  /** Get attachments for cell output items that have shouldDownloadImage=true */
  override async getAttachments(
    items: CellOutputContextItem[],
  ): Promise<FileUIPart[]> {
    const cellId = items[0].data.cellId;
    const cellName = items[0].data.cellName;
    const cellOutputs = items.map((item) => item.data.cellOutput);

    return getAttachmentsForOutputs(cellOutputs, cellId, cellName);
  }
}

export async function getAttachmentsForOutputs(
  data: BaseOutput[],
  cellId: CellId,
  cellName: string,
): Promise<FileUIPart[]> {
  // Filter items that need image downloading
  const itemsNeedingDownload = data.filter(
    (item) => item.shouldDownloadImage && item.outputType === "media",
  );

  if (itemsNeedingDownload.length === 0) {
    return [];
  }

  // Prepare download requests
  const downloadRequests = itemsNeedingDownload.flatMap((item) => {
    const outputElement = document.getElementById(CellOutputId.create(cellId));
    if (!outputElement) {
      Logger.warn(`Output element not found for cell ${cellId}`);
      return [];
    }
    return {
      cellId: cellId,
      cellName: cellName,
      mimetype: item.output.mimetype,
      element: outputElement,
    };
  });

  try {
    return await Promise.all(
      downloadRequests.map(async (item) => ({
        type: "file",
        filename: `${cellName}-output-screenshot`,
        mediaType: "image/png",
        url: await toPng(item.element),
      })),
    );
  } catch (error) {
    Logger.error("Error downloading cell output images:", error);
    return [];
  }
}

function getBaseOutput(output: OutputMessage): BaseOutput | null {
  if (isOutputEmpty(output)) {
    return null;
  }

  const mimetype = output.mimetype;

  // Determine output type
  const isMedia = isMediaMimetype(mimetype, String(output.data));
  const outputType = isMedia ? "media" : "text";

  let processedContent: string | undefined;
  let imageUrl: string | undefined;
  let shouldDownloadImage = false;

  // Process text content
  if (outputType === "text" && typeof output.data === "string") {
    processedContent =
      mimetype === "text/html" ? parseHtmlContent(output.data) : output.data;
  }

  // Process media content - for now, we'll just note that it's media
  if (outputType === "media") {
    if (typeof output.data === "string" && output.data.startsWith("data:")) {
      imageUrl = output.data; // Data URL
    } else if (
      typeof output.data === "string" &&
      output.data.startsWith("http")
    ) {
      imageUrl = output.data; // External URL
    } else {
      // Download the image from the DOM element
      shouldDownloadImage = true;
    }
  }

  return {
    processedContent,
    imageUrl,
    shouldDownloadImage,
    outputType,
    output,
  };
}

export function getCellContextData(
  cellId: CellId,
  notebook: NotebookState,
  opts?: {
    includeConsoleOutput: boolean;
  },
): CellOutputData {
  const { includeConsoleOutput = false } = opts || {};
  const cellRuntime = notebook.cellRuntime[cellId];

  const cellData = notebook.cellData[cellId];
  const cellIndex = notebook.cellIds.inOrderIds.indexOf(cellId);
  const cellName = displayCellName(cellData.name, cellIndex);

  let consoleOutputs: BaseOutput[] | undefined;
  if (includeConsoleOutput) {
    consoleOutputs = cellRuntime.consoleOutputs
      .map((output) => getBaseOutput(output))
      .filter((output) => output !== null);
  }

  let cellOutput: BaseOutput | undefined;
  const rawCellOutput = cellRuntime.output;
  if (rawCellOutput && !isOutputEmpty(rawCellOutput)) {
    cellOutput = getBaseOutput(rawCellOutput) ?? undefined;
  }

  return {
    cellId,
    cellName,
    cellCode: cellData.code,
    cellOutput,
    consoleOutputs,
  };
}
