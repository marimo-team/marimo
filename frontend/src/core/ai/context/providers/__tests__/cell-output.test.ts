/* Copyright 2024 Marimo. All rights reserved. */

import { beforeEach, describe, expect, it, vi } from "vitest";
import { Mocks } from "@/__mocks__/common";

// Mock the external dependencies
vi.mock("html-to-image", () => ({
  toPng: vi.fn().mockResolvedValue("data:image/png;base64,mockbase64data"),
}));

vi.mock("@/utils/Logger", () => ({
  Logger: Mocks.quietLogger(),
}));

import type { NotebookState } from "@/core/cells/cells";
import type { CellId } from "@/core/cells/ids";
import type { OutputMessage } from "@/core/kernel/messages";
import type { JotaiStore } from "@/core/state/jotai";
import { CellOutputContextProvider } from "../cell-output";

// Test helper to create mock store
function createMockStore(notebook: NotebookState): JotaiStore {
  const store = {
    get: vi.fn().mockReturnValue(notebook),
  } as unknown as JotaiStore;
  return store;
}

describe("CellOutputContextProvider", () => {
  let provider: CellOutputContextProvider;
  let mockStore: JotaiStore;
  let mockNotebook: NotebookState;

  beforeEach(() => {
    // Create a basic mock notebook state
    mockNotebook = {
      cellIds: {
        inOrderIds: ["cell1" as CellId, "cell2" as CellId, "cell3" as CellId],
      },
      cellData: {
        cell1: {
          id: "cell1" as CellId,
          name: "My Cell",
          code: "print('hello world')",
        },
        cell2: {
          id: "cell2" as CellId,
          name: "",
          code: "import matplotlib.pyplot as plt\nplt.plot([1,2,3])",
        },
        cell3: {
          id: "cell3" as CellId,
          name: "Empty Cell",
          code: "# no output",
        },
      },
      cellRuntime: {
        cell1: {
          output: {
            mimetype: "text/plain",
            data: "hello world",
          } as OutputMessage,
        },
        cell2: {
          output: {
            mimetype: "image/png",
            data: "base64imagedata",
          } as OutputMessage,
        },
        cell3: {
          output: null,
        },
      },
    } as unknown as NotebookState;

    mockStore = createMockStore(mockNotebook);
    provider = new CellOutputContextProvider(mockStore);
  });

  describe("basic properties", () => {
    it("should have correct title and context type", () => {
      expect(provider.title).toBe("Cell Outputs");
      expect(provider.contextType).toBe("cell-output");
      expect(provider.mentionPrefix).toBe("@");
    });
  });

  describe("getItems", () => {
    it("should return only cells with outputs", () => {
      const items = provider.getItems();

      // Should have 2 items (cell1 and cell2, but not cell3 which has no output)
      expect(items).toHaveLength(2);

      const cellIds = items.map((item) => item.data.cellId);
      expect(cellIds).toContain("cell1");
      expect(cellIds).toContain("cell2");
      expect(cellIds).not.toContain("cell3");
    });

    it("should correctly identify text vs media outputs", () => {
      const items = provider.getItems();

      const textItem = items.find((item) => item.data.cellId === "cell1");
      const mediaItem = items.find((item) => item.data.cellId === "cell2");

      expect(textItem?.data.outputType).toBe("text");
      expect(mediaItem?.data.outputType).toBe("media");
    });

    it("should process text content correctly", () => {
      const items = provider.getItems();
      const textItem = items.find((item) => item.data.cellId === "cell1");

      expect(textItem?.data.processedContent).toBe("hello world");
    });

    it("should mark media items for download when no direct URL available", () => {
      const items = provider.getItems();
      const mediaItem = items.find((item) => item.data.cellId === "cell2");

      expect(mediaItem?.data.shouldDownloadImage).toBe(true);
      expect(mediaItem?.data.imageUrl).toBeUndefined();
    });

    it("should include cell code and names", () => {
      const items = provider.getItems();

      const item1 = items.find((item) => item.data.cellId === "cell1");
      const item2 = items.find((item) => item.data.cellId === "cell2");

      expect(item1?.data.cellName).toBe("My Cell");
      expect(item1?.data.cellCode).toBe("print('hello world')");

      expect(item2?.data.cellName).toBe("cell-1");
      expect(item2?.data.cellCode).toBe(
        "import matplotlib.pyplot as plt\nplt.plot([1,2,3])",
      );
    });
  });

  describe("formatContext", () => {
    it("should format text output context correctly", () => {
      const items = provider.getItems();
      const textItem = items.find((item) => item.data.cellId === "cell1");

      if (!textItem) {
        throw new Error("Text item not found");
      }

      const context = provider.formatContext(textItem);

      expect(context).toContain("cell-output");
      expect(context).toContain("My Cell");
      expect(context).toContain("Cell Code:");
      expect(context).toContain("print('hello world')");
      expect(context).toContain("Output:");
      expect(context).toContain("hello world");
    });

    it("should format media output context correctly", () => {
      const items = provider.getItems();
      const mediaItem = items.find((item) => item.data.cellId === "cell2");

      if (!mediaItem) {
        throw new Error("Media item not found");
      }

      const context = provider.formatContext(mediaItem);

      expect(context).toContain("cell-output");
      expect(context).toContain("cell-1");
      expect(context).toContain("Cell Code:");
      expect(context).toContain("import matplotlib.pyplot as plt");
      expect(context).toContain("Media Output: Contains image/png content");
    });
  });

  describe("formatCompletion", () => {
    it("should create proper completion object", () => {
      const items = provider.getItems();
      const item = items[0];

      const completion = provider.formatCompletion(item);

      expect(completion.label).toMatch(/@.+/);
      expect(completion.displayLabel).toBe(item.data.cellName);
      expect(completion.detail).toContain("output");
      expect(completion.type).toBe("cell-output");
      expect(completion.section).toBe("Cell Output");
      expect(typeof completion.info).toBe("function");
    });
  });
});

// Test utility functions separately
describe("Cell output utility functions", () => {
  // We need to extract these functions to test them, or make them exported
  // For now, let's test through the public interface

  describe("media type detection", () => {
    let provider: CellOutputContextProvider;
    let mockStore: JotaiStore;

    beforeEach(() => {
      const mockNotebook = {
        cellIds: { inOrderIds: [] },
        cellData: {},
        cellRuntime: {},
      } as unknown as NotebookState;

      mockStore = createMockStore(mockNotebook);
      provider = new CellOutputContextProvider(mockStore);
    });

    it("should detect image mimetypes as media", () => {
      const testCases = [
        { mimetype: "image/png", data: "test", expected: "media" },
        { mimetype: "image/jpeg", data: "test", expected: "media" },
        { mimetype: "image/gif", data: "test", expected: "media" },
        { mimetype: "image/svg+xml", data: "test", expected: "media" },
      ];

      for (const testCase of testCases) {
        mockStore = createMockStore({
          cellIds: { inOrderIds: ["test" as CellId] },
          cellData: {
            test: { id: "test" as CellId, name: "", code: "" },
          },
          cellRuntime: {
            test: {
              output: {
                mimetype: testCase.mimetype,
                data: testCase.data,
              } as OutputMessage,
            },
          },
        } as unknown as NotebookState);

        provider = new CellOutputContextProvider(mockStore);
        const items = provider.getItems();

        expect(items[0]?.data.outputType).toBe(testCase.expected);
      }
    });

    it("should detect HTML with media tags as media", () => {
      const htmlWithImage = '<div><img src="test.png" /></div>';
      const htmlWithCanvas = '<canvas width="100" height="100"></canvas>';
      const htmlWithSvg = '<svg><circle r="50" /></svg>';

      const testCases = [
        { data: htmlWithImage, expected: "media" },
        { data: htmlWithCanvas, expected: "media" },
        { data: htmlWithSvg, expected: "media" },
        { data: "<p>Just text</p>", expected: "text" },
      ];

      for (const testCase of testCases) {
        mockStore = createMockStore({
          cellIds: { inOrderIds: ["test" as CellId] },
          cellData: {
            test: { id: "test" as CellId, name: "", code: "" },
          },
          cellRuntime: {
            test: {
              output: {
                mimetype: "text/html",
                data: testCase.data,
              } as OutputMessage,
            },
          },
        } as unknown as NotebookState);

        provider = new CellOutputContextProvider(mockStore);
        const items = provider.getItems();

        expect(items[0]?.data.outputType).toBe(testCase.expected);
      }
    });

    it("should detect text mimetypes correctly", () => {
      const testCases = [
        { mimetype: "text/plain", expected: "text" },
        {
          mimetype: "text/html",
          data: "<p>No media tags</p>",
          expected: "text",
        },
        { mimetype: "application/json", expected: "text" },
      ];

      for (const testCase of testCases) {
        mockStore = createMockStore({
          cellIds: { inOrderIds: ["test" as CellId] },
          cellData: {
            test: { id: "test" as CellId, name: "", code: "" },
          },
          cellRuntime: {
            test: {
              output: {
                mimetype: testCase.mimetype,
                data: testCase.data || "test data",
              } as OutputMessage,
            },
          },
        } as unknown as NotebookState);

        provider = new CellOutputContextProvider(mockStore);
        const items = provider.getItems();

        expect(items[0]?.data.outputType).toBe(testCase.expected);
      }
    });
  });

  describe("HTML content parsing", () => {
    let provider: CellOutputContextProvider;
    let mockStore: JotaiStore;

    beforeEach(() => {
      // Mock DOM methods for HTML parsing
      const mockDiv = {
        innerHTML: "",
        textContent: "",
        innerText: "",
      };

      global.document = {
        createElement: vi.fn().mockReturnValue(mockDiv),
      } as unknown as Document;

      const mockNotebook = {
        cellIds: { inOrderIds: [] },
        cellData: {},
        cellRuntime: {},
      } as unknown as NotebookState;

      mockStore = createMockStore(mockNotebook);
      provider = new CellOutputContextProvider(mockStore);
    });

    it("should extract text content from HTML", () => {
      const htmlContent = "<p>Hello <strong>world</strong>!</p>";

      // Mock the DOM parsing result
      const mockDiv = {
        innerHTML: "",
        textContent: "Hello world!",
        innerText: "Hello world!",
      };

      (
        global.document.createElement as ReturnType<typeof vi.fn>
      ).mockReturnValue(mockDiv);

      mockStore = createMockStore({
        cellIds: { inOrderIds: ["test" as CellId] },
        cellData: {
          test: { id: "test" as CellId, name: "", code: "" },
        },
        cellRuntime: {
          test: {
            output: {
              mimetype: "text/html",
              data: htmlContent,
            } as OutputMessage,
          },
        },
      } as unknown as NotebookState);

      provider = new CellOutputContextProvider(mockStore);
      const items = provider.getItems();

      expect(items[0]?.data.processedContent).toBe("Hello world!");
    });
  });
});
