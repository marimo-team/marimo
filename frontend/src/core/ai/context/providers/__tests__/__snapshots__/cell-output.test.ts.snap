// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CellOutputContextProvider > formatContext > should format media output context correctly > media-output-context 1`] = `
"<cell-output name="cell-1" cellId="cell2" outputType="media" mimetype="image/png">Cell Code:
import matplotlib.pyplot as plt
plt.plot([1,2,3])

Media Output: Contains image/png content</cell-output>"
`;

exports[`CellOutputContextProvider > formatContext > should format text output context correctly > text-output-context 1`] = `
"<cell-output name="My Cell" cellId="cell1" outputType="text" mimetype="text/plain">Cell Code:
print('hello world')

Output:
hello world</cell-output>"
`;
